src/java/org/jivesoftware/util/S2STestService.java

```diff
@@ -0,0 +1,213 @@
+package org.jivesoftware.util;
+
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+import javax.xml.bind.DatatypeConverter;
+
+import org.apache.log4j.Appender;
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+import org.apache.log4j.PatternLayout;
+import org.apache.log4j.WriterAppender;
+import org.apache.log4j.spi.LoggingEvent;
+import org.jivesoftware.openfire.XMPPServer;
+import org.jivesoftware.openfire.handler.IQPingHandler;
+import org.jivesoftware.openfire.interceptor.InterceptorManager;
+import org.jivesoftware.openfire.interceptor.PacketInterceptor;
+import org.jivesoftware.openfire.interceptor.PacketRejectedException;
+import org.jivesoftware.openfire.session.OutgoingServerSession;
+import org.jivesoftware.openfire.session.Session;
+import org.slf4j.LoggerFactory;
+import org.xmpp.packet.IQ;
+import org.xmpp.packet.IQ.Type;
+import org.xmpp.packet.JID;
+import org.xmpp.packet.Packet;
+
+/**
+ * Runs server to server test.
+ *
+ * Attempts to send an IQ packet to ping a given domain. Captures debug information from logging, certificates and
+ * packets.
+ */
+public class S2STestService {
+
+	private static final org.slf4j.Logger Log = LoggerFactory.getLogger(S2STestService.class);
+
+	private Semaphore waitUntil;
+
+	private String domain;
+
+	/**
+	 * @param domain The host to test.
+	 */
+	public S2STestService(String domain) {
+		this.domain = domain;
+	}
+
+	/**
+	 * Run a test against the domain.
+	 * @return K-V pairs of debug information.
+	 * @throws Exception On error.
+	 */
+	public Map<String, String> run() throws Exception {
+		waitUntil = new Semaphore(0);
+		Map<String, String> results = new HashMap<>();
+
+		// Tear down existing routes.
+		XMPPServer.getInstance().getRoutingTable().removeServerRoute(new JID(domain));
+
+		// Intercept logging.
+		final StringBuilder logs = new StringBuilder();
+		Appender appender = interceptLogging(logs);
+
+		// Intercept packets.
+		PacketInterceptor interceptor = new S2SInterceptor();
+		InterceptorManager.getInstance().addInterceptor(interceptor);
+
+		// Send ping.
+		Log.info("Sending server to server ping request.");
```

Comment:
could the server being sent to please be denoted in the log message, if not already via some magic I am unaware of :)